theano/tensor/extra_ops.py

```diff
@@ -11,6 +11,43 @@
 tensor = basic
 
 
+class cpu_contiguous(theano.Op):
+    def __eq__(self, other):
+        return type(self) == type(other)
+
+    def __hash__(self, other):
+        return hash(type(self))
+    
+    def make_node(self, x):
+        x_ = theano.tensor.as_tensor_variable(x)
+        return theano.Apply(self, [x_], [x_.type()])
+    
+    def perform(self, node, inputs, output_storage):
+        x, = inputs
+        y = output_storage[0][0]
+        # if the ouput is contiguous do nothing
+        # else return a copy
+        if not x.is_c_contiguous():
+            x = x.copy()
+        assert x.is_c_contiguous()
+        y = x 
+
+    def c_code(self, node, name, inames, onames, sub): 
+        x, = inames
+        y, = onames
+        # check if output is continguous, if it is copy it in.
+        code = """
+            if (PyArrayCHKFLAGS(%(y)s, NPY_ARRAY_C_CONTIGUOUS){
+                PyArray_CopyInto(%(y)s, %(x)s)
```

Comment:
If the input is contiguous, do not copy it. The output isn't preallocated by default.